### PR TITLE
wkw centralization

### PR DESCRIPTION
**Pull request by Aminelouati**
Let us be ready for the "Hub":
-Consider proactive-examples sub dir as package dirs
-Split load-examples.groovy into load-packages.groovy and LoadPackage.groovy

### CODE REVIEW

**Comment by tobwiens** [comment]
I cannot see the difference between the files. Github shows the whole file as "re-written" can you fix that? Or can you describe what you changed?

File: config/scheduler/settings.ini [hunk 1]
```diff
@@ -83,7 +83,7 @@ pa.scheduler.core.starttask.threadnumber=5
 pa.scheduler.core.listener.threadnumber=5
 
 # List of the scripts paths to execute at scheduler start. Paths are separated by a ';'.
-pa.scheduler.startscripts.paths=tools/load-examples.groovy
+pa.scheduler.startscripts.paths=tools/load-packages.groovy
```
Comment (mbenguig) [inline config/scheduler/settings.ini]: Why not calling the script "LoadPackages.groovy" ?

File: tools/LoadPackage.groovy [hunk 2]
```diff
@@ -0,0 +1,252 @@
+import org.ow2.proactive.authentication.ConnectionInfo
+import org.ow2.proactive.scheduler.rest.SchedulerClient
+import java.nio.file.Files
+import java.nio.file.Paths
+import java.nio.file.StandardCopyOption
+import org.apache.log4j.Logger;
+
+
+class LoadPackage {
```
Comment (mbenguig) [inline tools/LoadPackage.groovy]: (tested as a start script of the scheduler?)

File: tools/load-packages.groovy [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+class LoadPackages {
+
+
+    private final String SCRIPT_NAME = "load-packages.groovy"
+
+    private final String LOAD_PACKAGE_SCRIPT_NAME = "LoadPackage.groovy"
+
+    private final String BUCKET_OWNER = "GROUP:public-objects"
+
+    private final String GLOBAL_SPACE_PATH
```
Comment (mbenguig) [inline tools/load-packages.groovy]: Could you remove all useless variables?

File: tools/LoadPackage.groovy [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                    studio_metadata_file.text = studio_template.get("offsets_json_string")
+                    writeToOutput(" " + studio_metadata_file.absolutePath + " created!")
+
+                    template_dir_name = (template_dir_name.toInteger() + 1) + ""
+                }
+
+            }
+
+        }
+    }
```
Comment (mbenguig) [inline tools/LoadPackage.groovy]: The user must be able to load packages independently calling this script... and without necessarily being called from SchedulerStarter (ie without bindings).. Cf Brian comments on the related trello card for the explanations

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 646 issues
* CRITICAL 1 critical
* MAJOR 3 major
* MINOR 642 minor

#### Top 10 issues

1. CRITICAL [InstallPackageCommand.java#L124](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java#L124): Use a logger to log this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1148)
1. MAJOR [InstallPackageCommand.java#L133](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java#L133): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [SchedulerClient.java#L1212](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java#L1212): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [SchedulerClient.java#L1214](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java#L1214): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MINOR [Scheduler.java#L83](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L83): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L92](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L92): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L114](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L114): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L157](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L157): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L187](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L187): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L231](https://github.com/ow2-proactive/scheduling/blob/6943a8afb799803990d3032238d31813932766f2/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L231): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            }
+            // Scripts binding
+            schedulerProperties.put("package.path.name", PACKAGE_PATH_NAME);
+            schedulerProperties.put("session.id", currentContext.getSessionId());
+            ScriptHandler scriptHandler = new ScriptHandler();
+            scriptHandler.addBindings(schedulerProperties);
+
+            // Execute the script
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            PrintStream ps = new PrintStream(os, true);
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: give the variables expressive names

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                                                     PACKAGE_PATH_NAME));
+            }
+            // Scripts binding
+            schedulerProperties.put("package.path.name", PACKAGE_PATH_NAME);
+            schedulerProperties.put("session.id", currentContext.getSessionId());
+            ScriptHandler scriptHandler = new ScriptHandler();
+            scriptHandler.addBindings(schedulerProperties);
+
+            // Execute the script
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: Make this variable name more expressive

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        this.PACKAGE_PATH_NAME = packagePathName;
+
+    }
+
+    @Override
+    public void execute(ApplicationContext currentContext) throws CLIException {
+
+        SchedulerRestInterface scheduler = currentContext.getRestClient().getScheduler();
+        try {
+            //retrieve the scheduler properties
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: Put this "retrieve the scheduler properties" into a private method or a new class

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            Map<String, Object> schedulerProperties = scheduler.getSchedulerPropertiesFromSessionId(currentContext.getSessionId());
+            resultStack(currentContext).push(schedulerProperties);
+            logger.info("The scheduler properties are retrieved");
+
+            // Retrieve the script path
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: Put this "Retrieve the script path" into a private method or a new class

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+            if (!validatePackagePath()) {
+                throw new CLIException(REASON_INVALID_ARGUMENTS,
+                                       String.format("'%s'does not exist or is not a valid Package path.",
+                                                     PACKAGE_PATH_NAME));
+            }
+            // Scripts binding
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: Put this "Scripts binding" into a private method or a new class

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            schedulerProperties.put("package.path.name", PACKAGE_PATH_NAME);
+            schedulerProperties.put("session.id", currentContext.getSessionId());
+            ScriptHandler scriptHandler = new ScriptHandler();
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: This doesn't seem to be testable. Can you put this into a new class or private method to make this code testable

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            ScriptResult scriptResult;
+            File scriptFile;
+            scriptFile = new File(PASchedulerProperties.getAbsolutePath(SCRIPT_PATH));
+            if (scriptFile.exists()) {
+                logger.info("Executing " + SCRIPT_PATH);
+                String[] param = { PACKAGE_PATH_NAME };
+                scriptResult = scriptHandler.handle(new SimpleScript(scriptFile, param), ps, ps);
+                if (scriptResult.errorOccured()) {
+
+                    // Close streams before throwing
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: Put the clsing of streams into the finally block

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+                                                     scriptResult.getException().getMessage(),
+                                                     scriptResult.getException());
+                }
+                logger.info(os.toString());
+                os.reset();
+            } else {
+                logger.warn("Start script " + SCRIPT_PATH + " not found");
+            }
+
+            // Close streams
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: Put the closing of streams into the finally block

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            e.printStackTrace();
+        }
+
+    }
+
+    private boolean validatePackagePath() {
+        File file = new File(PACKAGE_PATH_NAME);
+        try {
+            return ((file.exists() && file.isDirectory()) || (file.exists() && file.getPath().endsWith(".zip")));
+        } catch (Exception e) {
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: Don't use exceptions for control flow!

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+            // Execute the script
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            PrintStream ps = new PrintStream(os, true);
+            ScriptResult scriptResult;
+            File scriptFile;
+            scriptFile = new File(PASchedulerProperties.getAbsolutePath(SCRIPT_PATH));
+            if (scriptFile.exists()) {
+                logger.info("Executing " + SCRIPT_PATH);
+                String[] param = { PACKAGE_PATH_NAME };
+                scriptResult = scriptHandler.handle(new SimpleScript(scriptFile, param), ps, ps);
```
Comment (tobwiens) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java]: What script do you execute here?

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java [hunk 15]
```diff
@@ -3667,4 +3667,25 @@ public JobIdData copyAndResubmitWithGeneralInfo(@HeaderParam("sessionid") String
...
     }
 
+    @GET
+    @Override
+    @Path("properties")
+    @Produces("application/json")
+    public Map<String, Object> getSchedulerPropertiesFromSessionId(@HeaderParam("sessionid")
+    final String sessionId) throws NotConnectedRestException, PermissionRestException {
+        SchedulerProxyUserInterface scheduler = checkAccess(sessionId, "properties");
+        Map<String, Object> schedulerProperties = null;
```
Comment (tobwiens) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java]: Initialize this map. With an empty map.

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java [hunk 16]
```diff
@@ -3667,4 +3667,25 @@ public JobIdData copyAndResubmitWithGeneralInfo(@HeaderParam("sessionid") String
...
+        try {
+            schedulerProperties = scheduler.getSchedulerProperties();
+        } catch (NotConnectedException e) {
+            logger.warn("Attempt to retrieve scheduler properties but failed because connection exception", e);
+            return null;
```
Comment (tobwiens) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java]: Don't return null, forward the exception.

File: rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java [hunk 17]
```diff
@@ -3667,4 +3667,25 @@ public JobIdData copyAndResubmitWithGeneralInfo(@HeaderParam("sessionid") String
...
+        } catch (PermissionException e) {
+            logger.warn("Attempt to retrieve scheduler properties but failed because permission exception", e);
+            return null;
```
Comment (tobwiens) [inline rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java]: Don't return null, forward the exception.

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontendState.java [hunk 18]
```diff
@@ -1217,4 +1216,13 @@ public UserData getCurrentUserData() throws NotConnectedException {
         return users;
     }
 
+    public Map<String, Object> getSchedulerProperties() throws NotConnectedException {
+        UniqueID id = checkAccess();
+
+        UserIdentificationImpl ident = identifications.get(id).getUser();
+        // renew session for this user
```
Comment (tobwiens) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontendState.java]: This comment is redundant you can remove it

**Comment by tobwiens** [comment]
@Aminelouati 
- Can you add unit tests for all the code you have written. 
- Can you write systemtests which guard this functionality

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 647 issues
* MAJOR 3 major
* MINOR 644 minor

#### Top 10 issues

1. MAJOR [InstallPackageCommand.java#L112](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/cmd/sched/InstallPackageCommand.java#L112): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [SchedulerClient.java#L1212](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java#L1212): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MAJOR [SchedulerClient.java#L1214](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/rest/rest-client/src/main/java/org/ow2/proactive/scheduler/rest/SchedulerClient.java#L1214): Either log or rethrow this exception. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS1166)
1. MINOR [Scheduler.java#L83](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L83): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L92](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L92): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L114](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L114): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L157](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L157): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L187](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L187): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L231](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L231): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. MINOR [Scheduler.java#L267](https://github.com/ow2-proactive/scheduling/blob/182efd542eccc8c835e30419abf008c08532dda9/scheduler/scheduler-api/src/main/java/org/ow2/proactive/scheduler/common/Scheduler.java#L267): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

File: rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/RmCommandFactory.java [hunk 19]
```diff
@@ -25,12 +25,6 @@
  */
 package org.ow2.proactive_grid_cloud_portal.cli;
 
-import java.util.List;
```
Comment (mbenguig) [inline rest/rest-cli/src/main/java/org/ow2/proactive_grid_cloud_portal/cli/RmCommandFactory.java]: Useless Imports ?

File: tools/load-examples.groovy [hunk 20]
```diff
@@ -1,341 +0,0 @@
-import org.ow2.proactive.authentication.ConnectionInfo
```
Comment (fviale) [inline tools/load-examples.groovy]: ideally, we should see in the pull-request the renaming from load-examples.groovy, to LoadPackage.groovy, the PR does not show this as a renaming but as a delete file and new file, making it difficult to track the changes

**Comment by fviale** [comment]
This pull request is very big, it requires a better naming "than wkw centralization" and a more detailed explanation of its content in the PR description

**Pull request closed without merging**